### PR TITLE
Last value validation

### PR DESCRIPTION
**Pull request by VWSCoronaDashboard8**
## Summary

For every time series in data, try to match last_value with the last value in the values array. As part of this work I have unified the code around timestamps and data sorting to all use the same common functions. Also the type guards and logic have been improved in some places.

The script can be run with `yarn validate-last-values` and optionally had the flags -e (fail early, on first errors) and -v for verbose output.

`sortTimeSeriesInDataInPlace` has been made context agnostic, to replace functions like `sortMunicipalTimeSeriesInDataInPlace`

I encountered an odd exception around `nl_deceased_rivm_per_age_group_value` because it doesn't have a time stamp. Even though we don't currently use it in the code, I think it makes sense to just give this data a timestamp, since it does originate from some place in time, and it avoids writing exceptions like that (although now I've improved the code to auto-detect timestamped series). I have added the timestamp but in a backwards-compatible way so we don't have to wait for backend to implement it before merging.

### CODE REVIEW

File: packages/app/schema/nl/deceased_rivm_per_age_group.json [hunk 1]
```diff
@@ -15,8 +15,10 @@
   "definitions": {
     "value": {
       "title": "nl_deceased_rivm_per_age_group_value",
-      "additionalProperties": false,
+      "//1": "Set to false once BE added date_unix timestamp",
```
Comment (VWSCoronaDashboard10) [inline packages/app/schema/nl/deceased_rivm_per_age_group.json:18]: I think you should be able to include comments with the `$comments` keyword: https://github.com/json-schema-org/json-schema-spec/pull/359

File: packages/cli/src/validate-last-values/index.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  console.log('Fail early?', cli.flags.failEarly);
+
+  const files = getFilesWithTimeSeries(directory);
+
+  const allFailures: Failure[] = [];
+
+  for (const file of files) {
+    const data = readJsonFile(path.join(directory, file));
+
+    sortTimeSeriesInDataInPlace(data);
```
Comment (VWSCoronaDashboard10) [inline packages/cli/src/validate-last-values/index.ts:68]: I don't really like the mutation pattern used here, but that refactor should probably not be part of this PR.

File: packages/cli/src/validate-last-values/index.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+      const promisedOperations = timeSeriesDataArray.map((metricData) => {
+        const installationName = metricData.rwzi_awzi_name;
+
+        const success = validateLastValue(metricData);
+        return {
+          success,
+          metricName: `sewer_per_installation.${installationName}`,
+        };
+      });
```
Comment (VWSCoronaDashboard10) [inline packages/cli/src/validate-last-values/index.ts:119]: It doesn't look like the map is returning a promise

File: packages/cli/src/validate-last-values/logic.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  isDateSeries,
+  isDateSpanSeries,
+  isTimeSeries,
+} from '@corona-dashboard/common';
+
+export function getTimeSeriesMetricNames(object: UnknownObject) {
+  return Object.keys(object).filter((key) => isTimeSeries(object[key]));
+}
+
+export function readJsonFile(filePath: string): UnknownObject {
```
Comment (VWSCoronaDashboard10) [inline packages/cli/src/validate-last-values/logic.ts:16]: `UnknownObject` is a bit too narrow, parsed JSON could for example also be an array. Maybe add a generic type?

File: packages/common/src/data-sorting.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  return Object.entries(data).reduce(
+    (acc, [propertyKey, propertyValue]) =>
+      isTimeSeries(propertyValue) ? [...acc, propertyKey as keyof T] : acc,
+    [] as (keyof T)[]
+  );
+}
+
+export function sortTimeSeriesValues(values: TimestampedValue[]) {
+  /**
+   * There are 3 ways in which time series data can be timestamped. We need to
```
Comment (VWSCoronaDashboard10) [inline packages/common/src/data-sorting.ts:44]: you're handling 2 instead of 3

File: packages/common/src/utils/assert.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export function assert(condition: any, msg: string): asserts condition {
```
Comment (VWSCoronaDashboard10) [inline packages/common/src/utils/assert.ts:1]: You might want to add a TSDoc comment about the `condition` matching any falsy value (e.g. passing `0` would fail the assertion)

**Review by VWSCoronaDashboard10 — APPROVED** [review]

**Review by VWSCoronaDashboard4 — APPROVED** [review]
Neat refactor!

File: packages/cli/src/validate-last-values/index.ts (hunk 2 above)
Comment (VWSCoronaDashboard8) [reply packages/cli/src/validate-last-values/index.ts:68]: I know you don't like mutating 😄 But yeah definitely not going to do that as part of this PR. I don't have a problem with it in this case. I imaging not mutating will be more hassle than I think is worth it. All data is only mutated/touched once before getting used.

File: packages/cli/src/validate-last-values/index.ts (hunk 3 above)
Comment (VWSCoronaDashboard8) [reply packages/cli/src/validate-last-values/index.ts]: Good catch!

File: packages/cli/src/validate-last-values/logic.ts (hunk 4 above)
Comment (VWSCoronaDashboard8) [reply packages/cli/src/validate-last-values/logic.ts:16]: We don't read files with array as their root. Adding it in the types only makes us having to program around it. I'm not sure that it adds any value then?

File: packages/common/src/utils/assert.ts (hunk 6 above)
Comment (VWSCoronaDashboard8) [reply packages/common/src/utils/assert.ts:1]: It is standard for 0 to fail an assertion AFAIK

Comment (VWSCoronaDashboard8) [reply packages/common/src/utils/assert.ts:1]: I remember having used a second assert specifically for undefined like `assertDefined` but I think it's easier to have just 1 function and write `assert(isDefined(condition),...)`

Comment (VWSCoronaDashboard8) [reply packages/common/src/utils/assert.ts:1]: I will add a comment about falsy values.

File: packages/cli/src/validate-last-values/logic.ts (hunk 4 above)
Comment (VWSCoronaDashboard8) [reply packages/cli/src/validate-last-values/logic.ts:16]: Alternatively I could rename the function readObjectFromJsonFile ;) @VWSCoronaDashboard10

**Review by VWSCoronaDashboard1 — APPROVED** [review]
I would have prefered to see this implemented as a custom keyword in the JSON schema validation. Akin to the `equalsRootProperty` keyword. Granted, it would have been a little harder to share the sorting function because the AJV code is still JS. But it might have been worth porting that logic to TS and then implementing this as a keyword in order to keep all of the validation in one place instead of disparate scripts.
But this works too, so I won't make too big of a fuss over it.

**Comment by VWSCoronaDashboard8** [comment]
@VWSCoronaDashboard1 I didn't realize that was an option. I'm not familiar with the capabilities of AJV. I suspect it might have been tricky to validate sewer per installation that way.

**Comment by VWSCoronaDashboard1** [comment]
> @VWSCoronaDashboard1 I didn't realize that was an option. I'm not familiar with the capabilities of AJV. I suspect it might have been tricky to validate sewer per installation that way.

It would have been a bit trickier, but from what I can quickly glean, it wouldn't be impossible. I didn't notice in time you were working on this, otherwise, I could have discussed it beforehand with you. Probably not worth the re-write since there's plenty of other things still on our list.
